subsys/ipc/ipc_service/lib/icmsg.c

```diff
@@ -159,31 +162,47 @@ static void submit_work_if_buffer_free_and_data_available(
 
 static void mbox_callback_process(struct k_work *item)
 {
+	char *rx_buffer;
 	struct icmsg_data_t *dev_data = CONTAINER_OF(item, struct icmsg_data_t, mbox_work);
 
 	atomic_t state = atomic_get(&dev_data->state);
-	int len = spsc_pbuf_read(dev_data->rx_ib, dev_data->rx_buffer,
-				 RX_BUF_SIZE);
 
-	__ASSERT_NO_MSG(len <= RX_BUF_SIZE);
+	uint16_t len = spsc_pbuf_claim(dev_data->rx_ib, &rx_buffer);
 
-	if (len == -EAGAIN) {
-		__ASSERT_NO_MSG(false);
-		submit_mbox_work(dev_data);
-		return;
-	} else if (len <= 0) {
+	if (len == 0) {
+		/* Unlikely, no data in buffer. */
 		return;
 	}
 
 	if (state == ICMSG_STATE_READY) {
 		if (dev_data->cb->received) {
-			dev_data->cb->received(dev_data->rx_buffer, len,
+#if CONFIG_IPC_SERVICE_ICMSG_NOCOPY_RX
+			dev_data->rx_buffer = rx_buffer;
+			dev_data->rx_len = len;
+#endif
+
+			dev_data->cb->received(rx_buffer, len,
 					       dev_data->ctx);
+
+			/* Release Rx buffer here only in case when user did not request
+			 * to hold it.
+			 */
+			if (is_rx_buffer_free(dev_data)) {
```

Comment:
```suggestion
			if (!is_rx_buffer_held(dev_data)) {
```
would be easier to read

Reply:
Is it make sense to you to create the new static function for opposite condition checking?

Reply:
Yes, it improves code readability. You are interested here if the callback is requested to hold the buffer. Not if it is "free" whatever it means in this context. This change would clearly show the programmer's intention. While the assembly should be exactly the same after optimization.

Reply:
Ok, done